plasmapy/utils/decorators/checks.py

```diff
@@ -1,20 +1,1219 @@
+        allowed_units = []
+        for target in targets:
+            try:
+                target_unit = u.Unit(target)
+                allowed_units.append(target_unit)
+            except TypeError as err:
+                # not a unit type
+                if not from_annotations:
+                    raise err
+
+                continue
+
+        return allowed_units
+
+    @staticmethod
+    def _normalize_equivalencies(equivalencies):
+        """
+        Normalizes equivalencies to ensure each is in a 4-tuple form::
+
+            (from_unit, to_unit, forward_func, backward_func)
+
+        `forward_func` maps `from_unit` into `to_unit` and `backward_func` does
+        the reverse.
+
+        Parameters
+        ----------
+        equivalencies: list of equivalent pairs
+            list of astropy :mod:`~astropy.units.equivalencies` to be normalized
+
+        Raises
+        ------
+        ValueError
+            if an equivalency can not be interpreted
+
+        Notes
+        -----
+        * the code here was copied and modified from
+          :func:`astropy.units.core._normalize_equivalencies` from AstroPy
+          version 3.2.3
+        * this will work on both the old style list equivalencies (pre AstroPy v3.2.1)
+          and the modern equivalencies defined with the
+          :class:`~astropy.units.equivalencies.Equivalency` class
+        """
+        if equivalencies is None:
+            return []
+
+        normalized = []
+
+        for i, equiv in enumerate(equivalencies):
+            if len(equiv) == 2:
+                from_unit, to_unit = equiv
+                a = b = lambda x: x
+            elif len(equiv) == 3:
+                from_unit, to_unit, a = equiv
+                b = a
+            elif len(equiv) == 4:
+                from_unit, to_unit, a, b = equiv
+            else:
+                raise ValueError(
+                    f"Invalid equivalence entry {i}: {equiv!r}")
+
+            if not (from_unit is u.Unit(from_unit) and
+                    (to_unit is None or to_unit is u.Unit(to_unit)) and
+                    callable(a) and
+                    callable(b)):
+                raise ValueError(
+                    f"Invalid equivalence entry {i}: {equiv!r}")
+            normalized.append((from_unit, to_unit, a, b))
+
+        return normalized
+
+    def _flatten_equivalencies_list(self, elist):
+        """
+        Given a list of equivalencies, flatten out any sub-element lists
+
+
+        Parameters
+        ----------
+        elist: list
+            list of astropy :mod:`~astropy.units.equivalencies` to be flattened
+
+        Returns
+        -------
+        list
+            a flattened list of astropy :mod:`~astropy.units.equivalencies`
+
+        """
+        new_list = []
+        for el in elist:
+            if not isinstance(el, list):
+                new_list.append(el)
+            else:
+                new_list.extend(self._flatten_equivalencies_list(el))
```

Comment:
Ooooh, my recursion sense is tingling! This would deal with any level of nesting, if I'm correct, right?